### ISSUE TITLE
[question] Are there environment difference between the "developer workflow" and `conan create`?

### ISSUE THREAD

**Issue by Adnn** [conan-io/conan#3529]
While moving one of our _fat_ dependencies from prebuilt binaries (packaged through Conan ; ) to a recipe actually building the thing, we are encountering a complication we never encountered for our other recipes:

We wrote the recipe, following the recommendend [Package Development Flow](https://docs.conan.io/en/latest/developing_packages/package_dev_flow.html#package-development-flow). The end to end flow works without any error, notably `conan build` command.

Yet, when moving to the final step of calling `conan create` on the exact same recipe, the process fails during the `build()`  step.

Are there known difference between the environments used by the "developer workflow" and `conan create` command that we could investigate to find the source of this failure?

**Comment by Adnn** [comment]
### Gory details

Due to the nature of this problem appearing in the middle of a 3rd party "mega build", we cannot really isolate a minimum to quickly demonstrate the issue (hence the very vague question, sorry for that).
This problem is appearing while building Qt 5.9.6 on OSX 10.12, using Conan 1.7.1

The error when using `conan create` ends with:

<details>
<summary>(... and there is so much more spewing before that)</summary>
/Users/adn/.conan/data/qt/5.9.6/ag/testing/build/3cfade032d90e40671f831047a4b021865f8396f/qt-everywhere-opensource-src-5.9.6/qtwebengine/src/3rdparty/chromium/tools/gn/._builder_record.cc:1:190: error: expected unqualified-id
<U+0000><U+0005><U+0016><U+0007><U+0000><U+0002><U+0000><U+0000>Mac OS X        <U+0000><U+0002><U+0000><U+0000><U+0000>   <U+0000><U+0000><U+0000>2<U+0000><U+0000><U+0000><B0><U+0000><U+0000><U+0000><U+0002><U+0000><U+0000><U+0000><E2><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000>ATTR<U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><E2><U+0000><U+0000><U+0000><98><U+0000><U+0000><U+0000>J<U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0001><U+0000><U+0000><U+0000><98><U+0000><U+0000><U+0000>J<U+0000><U+0000><U+0015>com.apple.quarantine<U+0000>q/0081;5b8eaa09;Google\x20Chrome.app;057AC61C-1740-47C1-BFBF-345B35B9491F
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               ^
/Users/adn/.conan/data/qt/5.9.6/ag/testing/build/3cfade032d90e40671f831047a4b021865f8396f/qt-everywhere-opensource-src-5.9.6/qtwebengine/src/3rdparty/chromium/tools/gn/._builder_record.cc:1:226: warning: null character ignored [-Wnull-character]
<U+0000><U+0005><U+0016><U+0007><U+0000><U+0002><U+0000><U+0000>Mac OS X        <U+0000><U+0002><U+0000><U+0000><U+0000>   <U+0000><U+0000><U+0000>2<U+0000><U+0000><U+0000><B0><U+0000><U+0000><U+0000><U+0002><U+0000><U+0000><U+0000><E2><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000>ATTR<U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><E2><U+0000><U+0000><U+0000><98><U+0000><U+0000><U+0000>J<U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0001><U+0000><U+0000><U+0000><98><U+0000><U+0000><U+0000>J<U+0000><U+0000><U+0015>com.apple.quarantine<U+0000>q/0081;5b8eaa09;Google\x20Chrome.app;057AC61C-1740-47C1-BFBF-345B35B9491F<U+0000>
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ^
96 warnings and 10 errors generated.
[207/463] CXX base/message_loop/message_pump_mac.o
/Users/adn/.conan/data/qt/5.9.6/ag/testing/build/3cfade032d90e40671f831047a4b021865f8396f/qt-everywhere-opensource-src-5.9.6/qtwebengine/src/3rdparty/chromium/base/message_loop/message_pump_mac.mm:643:53: warning: 'NSAnyEventMask' is deprecated: first deprecated in macOS 10.12 [-Wdeprecated-declarations]
      NSEvent* event = [NSApp nextEventMatchingMask:NSAnyEventMask
                                                    ^~~~~~~~~~~~~~
                                                    NSEventMaskAny
/System/Library/Frameworks/AppKit.framework/Headers/NSEvent.h:158:26: note: 'NSAnyEventMask' has been explicitly marked deprecated here
static const NSEventMask NSAnyEventMask                 API_DEPRECATED_WITH_REPLACEMENT("NSEventMaskAny", macosx(10.0, 10.12)) = NSUIntegerMax;
                         ^
/Users/adn/.conan/data/qt/5.9.6/ag/testing/build/3cfade032d90e40671f831047a4b021865f8396f/qt-everywhere-opensource-src-5.9.6/qtwebengine/src/3rdparty/chromium/base/message_loop/message_pump_mac.mm:665:48: warning: 'NSApplicationDefined' is deprecated: first deprecated in macOS 10.12 [-Wdeprecated-declarations]
  [NSApp postEvent:[NSEvent otherEventWithType:NSApplicationDefined
                                               ^~~~~~~~~~~~~~~~~~~~
                                               NSEventTypeApplicationDefined
/System/Library/Frameworks/AppKit.framework/Headers/NSEvent.h:79:26: note: 'NSApplicationDefined' has been explicitly marked deprecated here
static const NSEventType NSApplicationDefined   API_DEPRECATED_WITH_REPLACEMENT("NSEventTypeApplicationDefined", macosx(10.0, 10.12)) = NSEventTypeApplicationDefined;
                         ^
2 warnings generated.
[210/463] CXX base/metrics/field_trial.o
ninja: build stopped: subcommand failed.
Command '['/Users/adn/.conan/data/qt/5.9.6/ag/testing/build/3cfade032d90e40671f831047a4b021865f8396f/qtwebengine/src/3rdparty/ninja/ninja', '-C', '/Users/adn/.conan/data/qt/5.9.6/ag/testing/build/3cfade032d90e40671f831047a4b021865f8396f/qtwebengine/src/3rdparty/chromium/tools/gn/out/Release', 'gn']' returned non-zero exit status 1
Project ERROR: GN build error!
make[3]: *** [sub-gn-pro-make_first] Error 3
make[2]: *** [sub-buildtools-make_first] Error 2
make[1]: *** [sub-src-make_first] Error 2
make: *** [module-qtwebengine-make_first] Error 2
qt/5.9.6@ag/testing:
qt/5.9.6@ag/testing: ERROR: Package '3cfade032d90e40671f831047a4b021865f8396f' build failed
qt/5.9.6@ag/testing: WARN: Build folder /Users/adn/.conan/data/qt/5.9.6/ag/testing/build/3cfade032d90e40671f831047a4b021865f8396f
ERROR: qt/5.9.6@ag/testing: Error in build() method, line 91
	self.run("make -j {cpu_count} -s".format(cpu_count=tools.cpu_count()))
	ConanException: Error 512 while executing make -j 8 -s
</details>

**Comment by lasote** [comment]
Do you have `test_package` conanfile? Could we see the generated `conanbuildinfo.txt` and `conaninfo.txt`? and the applied profile?

**Comment by Adnn** [comment]
Thank you @lasote for taking a look at this issue.

I have no `test_package` for this recipe yet. 

Do you want the files when calling `conan create` ?

Here is the (uninteresting) `conanbuildinfo.txt` found in the build folder.

<details>
<summary>conanbuildinfo.txt</summary>
<p>

```ini
[includedirs]


[libdirs]


[bindirs]


[resdirs]


[builddirs]


[libs]


[defines]


[cppflags]


[cflags]


[sharedlinkflags]


[exelinkflags]


[sysroot]
```
</p>
</details>

Is there a `conaninfo.txt` when calling `conan create`?

As for my (default) profile:

<details>
<summary>default</summary>
<p>

```ini
[settings]
os=Macos
os_build=Macos
arch=x86_64
arch_build=x86_64
compiler=apple-clang
compiler.version=9.0
compiler.libcxx=libc++
build_type=Release
[options]
[build_requires]
[env]
```
</p>
</details>

**Comment by lasote** [comment]
> Is there a conaninfo.txt when calling conan create?

Sorry I was interested in the files when you call `conan install` in the local development flow. But honestly, I don't think I'm going to find anything but just in case. We know there can be slight differences between the local methods and the create but I'm not sure if big enough to break your build like that. Are you using the `AutotoolsBuildEnvironment` or calling directly the command line?

**Comment by Adnn** [comment]
After long on-and-off periods of trying to understand the problem, I think I can finally provide a description of what happens. The actual problem is actually triggered by the combination of different elements: OS X [a], Qt(ninja) [b], and Conan.

# Reproducing the issue
 Note that the situation described should only be the same on **OS X platforms**:

Here is a repository that allows to illustrate the issue:
https://github.com/Adnn/issue_conan-3529

The trivial `conanfile.py`

```
from conans import ConanFile, tools

class Issue3529Conan(ConanFile):
    name = "issue3529"
    version = "0.0.1"
    description = "see: https://github.com/conan-io/conan/issues/3529"
    settings = "os", "compiler", "build_type", "arch"

    no_copy_source = False

    def source(self):
        tools.get("https://github.com/Adnn/issue_conan-3529/releases/download/0.0.1/qtxml.tar.bz2")
```

Let's clone it and move into the root of the repo
```
git clone https://github.com/Adnn/issue_conan-3529.git
cd issue_conan-3529/
```

### Developer workflow

```
mkdir source
conan source .  -sf=source
mkdir build
cp -r source/* build/ # we simulate a recipe that needs to copy the source
```

    ls -la source/qt-everywhere-opensource-src-5.9.6/qtxmlpatterns/

> drwxr-xr-x   3 adn  wheel  102  3 Oct 15:40 .
> drwxr-xr-x   3 adn  wheel  102  3 Oct 15:40 ..
> drwxr-xr-x  20 adn  wheel  680  3 Oct 15:40 qtxmlpatterns

    ls -la build/qt-everywhere-opensource-src-5.9.6/qtxmlpatterns/

> drwxr-xr-x   4 adnn  wheel   136  3 Oct 15:39 .
> drwxr-xr-x   3 adnn wheel   102  3 Oct 15:39 ..
> -rwxr-xr-x   1 adnn  wheel   226  3 Oct 10:58 ._qtxmlpatterns
> drwxr-xr-x  38 adnn  wheel  1292  3 Oct 10:58 qtxmlpatterns

Here is the difference: 
The source folder contains a `._` prefixed version of each file and folder it contains, whereas the build one does not.

### Normal (create) workflow

```
conan create . com/testing
ls -la ~/.conan/data/issue3529/0.0.1/com/testing/build/$(HASH)/qt-everywhere-opensource-src-5.9.6/
```

> drwxr-xr-x   4 adn  staff   136  3 Oct 15:51 .
> drwxr-xr-x   5 adn  staff   170  3 Oct 15:51 ..
> -rwxr-xr-x   1 adn  staff   226  3 Oct 10:58 ._qtxmlpatterns
> drwxr-xr-x  38 adn  staff  1292  3 Oct 10:58 qtxmlpatterns

With the `conan create` workflow, the source folder *that is copied inside the build folder* also contains the `._` prefixed version of the folder.

### Command line extraction

mkdir dl && cd dl
wget https://github.com/Adnn/issue_conan-3529/releases/download/0.0.1/qtxml.tar.bz2
tar -xf qtxml.tar.bz2
ls -la qt-everywhere-opensource-src-5.9.6/

> drwxr-xr-x   3 adn  wheel  102  3 Oct 15:56 .
> drwxr-xr-x   4 adn  wheel  136  3 Oct 15:56 ..
> drwxr-xr-x@ 20 adn  wheel  680  3 Oct 15:56 qtxmlpatterns

With command line `tar` extraction, *no* `._` file.

# Some explanations

Those extra  `._` file are storage for extended attributes, that is created by OS X builtin `tar` command (as well as others). See this [superuser entry](https://superuser.com/q/61185/375719).

Because of that, they would only exist for archives created on OS X[a].

When extracting those files with `tar`, they are of course not expanded (and used to apply the correct attributes to the actual file), as seen in the example above.

When retrieved with Conan's `tools.get()`, those are not properly handled, and the files are left around.
Now, in cases were those file would be present is source archives, their existence would never be noticed by the user (as I was fully unaware of them until today). Even if using `tools.get()`, those extra hidden files don't usually get in the way, and the compilation works just fine.

But it seems Qt [b] WebEngine module, which relies on *ninja* to build some parts, gets really thrown of by those files. That is why my initial Qt recipe fails to build: *ninja* trips on those `._` files that are present in the copied source folder.

---

Finally, back to the initial topic of this issue:

> why is there a difference in behaviour between between the "developer workflow" and `conan create`

Because, to fully reproduce the expected recipe that would need to copy the source folder to modify bits of it, I added a manual `cp` step to the developer workflow, *because I could not find any conan developer conan to do this copy otherwise*.

And the `cp` shipped with OS X took care of handling those `._` (applying the attributes, and not copying them over to destination).

So, here it is, a recipe that was working perfectly in "developer workflow" is breaking at build time with `conan create`.


# Potential directions from here

This issue took more time than I'd like to admit to understand, and I keep hope that there are potential things to do so others would not have to go through this.

Of course, one important solution is to not include those files in the archive, which seems possible by defining `COPYFILE_DISABLE` env var before calling `tar` (or creating the archives on a non-OS X machine).
Yet, it seems reasonable to assume that if people are not aware of it, they will unawarely produce such archives.

Another direction would be for `tools.get()` to handle those extra files at least for OS X systems. I don't know if Python implementations make any provision to help with that.

What would seem important to me though, would be to give the recipes developers the ability to call the "source copy logic" from a developer command (either `conan source` or `conan build` I guess). In my situation, if I was able use the same copy logic in my developer workflow than was implicitly used by `conan create`, I would probably have spent less time barking the wrong trees (as the developer workflow would have failed directly).

If you read up to there: congratulations !

**Comment by lasote** [comment]
Wow, many thanks for your investigation!
So, If I understood correctly:

- We cannot avoid that these files to exist within a tar file because the file could be compressed by a third party, right? like the `qtxml.tar.bz2`. So the `COPYFILE_DISABLE` var won't help us much.

- The goal is to unzip the file (by using the `tools.get`) but without generating these "._" files.

Am I right?

**Comment by Adnn** [comment]
Hi @lasote, and thanks a lot for taking time to look into that (despite a *lengthy* comment),

Your description seems right.Ideally, source archives should not contain those files, because they are expected to be platform-agnostic.

Yet as you pointed out, Conan has no control over archives produced by 3rd parties. (Just a personal gut-feeling: it seems to me that a non-negligible part of developers working on OS X are unaware of this `tar` peculiarity on the platform).
The `COPYFILE_DISABLE` variable would allow to produce archives files without those `._` files in them, but once the archive is produced it cannot help.

* A potential direction would indeed be for `tools.get` to take care of those `._`.  On OS X, that should always be a good policy: if an archive is expected to work with the builtin `tar` command, it means these `._` are not expected to exist in the expanded destination anyway.
Yet, if someone on another OS decides that he wants to actually store (and retrieve) files prefixed with `._` in the archive, that might break. This is unlikely of course, but to be aware of.

* Should Conan instead issue **warning** on the CLI that those files are present in the archives, and that they go against "platform-agnostic" good practices? This way, someone in my current situation (responsible for both archiving the sources and developing the recipe) would be informed that he should clean up the archives. Yet, it could be an issue for some recipe developer who rely on an already existing archive he cannot control.

* Should there be a new option in the recipe with 3 values:
1. warn(default): to just issue warning if the files exist, so by default people are encouraged to produce "clean" archives
2. discard: the `._` file are correctly handled on OSX, and simply discarded on other OSes
3. keep: the `._` files are extracted normally (the current behaviour)

---

And some related remark, which could close the only gap I found between `conan create` and the *developer workflow*:

Do you think it would be possible to add a `--copy ` flag to `conan build` command?
It would run Conan's logic responsible for copying the content of source folder over to the build folder, like in a call to `conan create` with the default `no_copy_source=False`.

**Comment by lasote** [comment]
To summarize, investigate if we should remove (or not extract) the `._ ` osx files from the downloaded tgz`s.

**Comment by jgsogo** [comment]
I've been reading around and those files are intended to store only metadata about associated files. They are created and deleted on the fly (creating tar file, copying files,...) depending on the underlying storage, if it cannot store this metadata those files will be there. All the comments agree that those files are just metadata and can be deleted (Macos provides a command for it: `dot_clean`, and also the env variable you0ve talked about).

People agree that **all** the `._*` files can be deleted, but I think we can be more conservative and delete just the files that matches another one with the same name in the same directory (just in case). So a function following these lines would be enough:

```python
def dot_clean(folder):
    files = os.listdir(folder)
    for f in files:
        full_name = os.path.join(folder, f)
        if os.path.isdir(full_name):
            dot_clean(full_name)
        elif f.startswith("._"):
            if os.path.exists(os.path.join(folder, f[2:])):
                os.remove(full_name)
```

I would delete these files after extraction and also in the local cache after copying sources, to check that nobody is using them inside the recipe during creation. And I wouldn't provide an opt-out (do you think someone will ever need it?)

----
There are also some other _weird_ files associated to Mac like `.DS_Store` and `.Trashes` (dir). Also, in other osys we can find things like `Thumbs.db`,... shall we take care of all of them too?

**Comment by jgsogo** [comment]
Hi @Adnn,

We've been trying to reproduce the issue using the Conan workflow and we have realized that the metadata related to Macos files is not copied neither compressed into the packages created by Conan itself. I think that this behavior is taken into account in the test I've written [here](https://github.com/conan-io/conan/pull/4103/files#diff-e889b04c499e74fe445b41deb19529d8R17).

So, it seems that your problem arises because you are using external zipped files (not created by Conan). We will be providing a `conans.tools.dot_clean` tool that you will be able to call in order to get rid of all those `._` files inside a directory, but we are not going to remove files on our side by default as we think that it could be dangerous.

Will this new tool be useful (and enough) to your use case? Thanks.

**Comment by Adnn** [comment]
@jgsogo Thank you for your prompte answer!

For the second point, I might be missing something, but I think I would feel really good with explicit copy request in dev. workflow (leaving the default behaviour as it is).

**Comment by jgsogo** [comment]
Let me rephrase it. We can provide an argument to source command, something like `conan source --copy-like-in-cache ...` but as you were afraid of the developer being unaware (or forgetting) the `dot_clean` function, the same could happen for this argument.
